pkg/distributor/ha_tracker.go

```diff
@@ -0,0 +1,316 @@
+	electedLock sync.RWMutex
+	elected     map[string]replicaTracker
+	done        chan struct{}
+	quit        context.CancelFunc
+	// The labels to look for as bytes.
+	clusterBytes []byte
+	replicaBytes []byte
+	prefix       string
+
+	// We should only update the timestamp this often.
+	writeTimeout int64
+
+	consulLock       sync.RWMutex
+	overwriteTimeout int64 // Timeout that has to occur before we overwrite to a different replica.
+	client           ring.KVClient
+}
+
+// HATrackerConfig contains the configuration require to
+// create a HA Tracker.
+type HATrackerConfig struct {
+	ReplicaLabel string
+	ClusterLabel string
+
+	KVStore ring.KVConfig
+}
+
+// RegisterFlags adds the flags required to config this to the given FlagSet
+func (cfg *HATrackerConfig) RegisterFlags(f *flag.FlagSet) {
+	f.StringVar(&cfg.ReplicaLabel, "ha-tracker.replica", "replica", "Prometheus label to look for in samples to identify a Proemtheus HA replica.")
+	f.StringVar(&cfg.ClusterLabel, "ha-tracker.cluster", "cluster", "Prometheus label to look for in samples to identify a Poemtheus HA cluster.")
+	// We want the ability to use different Consul instances for the ring and for HA cluster tracking.
+	cfg.KVStore.RegisterFlags(f, "ha-tracker")
+}
+
+// NewClusterTracker takes local as true to use local in memory KVStore
+// rather than consul.
+func NewClusterTracker(local bool, cfg HATrackerConfig) (clusterTracker, error) {
+	var client ring.KVClient
+	var err error
+
+	codec := ring.ProtoCodec{Factory: ProtoReplicaDescFactory}
+
+	// TODO: Find a nicer way to get this value into tracker.loop for WatchPrefix.
+	// Since we need to set a prefix for HA tracking, but passing a Consul config with
+	// a prefix creates a ring.PrefixClient, we save the prefix value and set it to
+	// empty string on the config.
+	prefix := cfg.KVStore.Consul.Prefix
+	cfg.KVStore.Consul.Prefix = ""
+
+	switch cfg.KVStore.Store {
+	case "consul":
+		client, err = ring.NewConsulClient(cfg.KVStore.Consul, codec)
+		if err != nil {
+			return nil, err
+		}
+	case "inmemory":
+		// TODO: do we want to use the same init do for sync.Once as we do in the ring?
+		client = ring.NewInMemoryKVClient(codec)
+	default:
+		return nil, fmt.Errorf("invalid KV store type: %s", cfg.KVStore.Store)
+	}
+
+	ctx, quit := context.WithCancel(context.Background())
+	t := &tracker{
+		logger:           util.Logger,
+		elected:          map[string]replicaTracker{},
+		writeTimeout:     10 * time.Second.Nanoseconds(),
+		overwriteTimeout: time.Minute.Nanoseconds(),
+		clusterBytes:     []byte(cfg.ClusterLabel),
+		replicaBytes:     []byte(cfg.ReplicaLabel),
+		prefix:           prefix,
+		client:           client,
+		quit:             quit,
+	}
+	go t.loop(ctx)
+	return t, nil
+}
+
+// Follows pattern used by ring for WatchKey.
+func (c *tracker) loop(ctx context.Context) {
+	defer close(c.done)
+	c.client.WatchPrefix(context.Background(), fmt.Sprintf("%s", c.prefix), func(value interface{}) bool {
```

Comment:
Why is `loop` in the interface?

And if you've going to take a ctx, you'll want to use it...

Also the sprintf is a noop.